drivers: usb_dc_mcux: Enable second USB controller on RT socs

danieldegrasse opened the pull request:
Enable USB device controller to handle either `usb1` or `usb2` nodes being enabled. Note that the USB driver is still a single instance driver, but it can now use either USB instance.  This PR also enables support for USB on the RT1160 and RT1170 evaluation kits, which was previously only enabled on RT10xx EVKs. USB support was tested using the mass storage and console sample.

danieldegrasse committed:
dts: rt11xx: Update RT11xx SOC DTS with correct USB clock device

RT11xx USB PLLs source from 24MHz XTAL oscillator. Add this oscillator
as a clock source for the usb dts entries for the RT11xx.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: rt11xx: enable usb pll clock

USB pll clock should be enabled if USB driver is used. Add required
configuration for USB pll clock.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1170_evk_cm7: Enable USB

USB controller was tested with usb console sample, and usb mass storage
sample.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1160_evk_cm7: Enable USB

Enable usb controller on rt1160 cm7. Tested using usb console sample.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented:
DNM until I've had a chance to review (I'm doing it now) please

mbolivar-nordic commented on soc/arm/nxp_imx/rt/soc_rt11xx.c:
```diff
@@ -409,6 +409,14 @@ static ALWAYS_INLINE void clock_init(void)
 #endif
 
+#if CONFIG_USB_DC_NXP_EHCI
+	CLOCK_EnableUsbhs0PhyPllClock(kCLOCK_Usb480M,
+		DT_PROP_BY_PHANDLE(DT_INST(0, nxp_mcux_usbd), clocks, clock_frequency));
+	CLOCK_EnableUsbhs0Clock(kCLOCK_Usb480M,
+		DT_PROP_BY_PHANDLE(DT_INST(0, nxp_mcux_usbd), clocks, clock_frequency));
+	USB_EhciPhyInit(kUSB_ControllerEhci0, CPU_XTAL_CLK_HZ, &usbPhyConfig);
```
This looks incorrect to me and I didn't find enough context in the commit logs to justify why it is correct.

In the previous patch, there are two nodes with compatible "nxp,mcux-usbd" in the DTS. And in general, you cannot make any assumptions about which one of them is instance 0. See https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_INST for details, especially the part beginning with "No other guarantees are made. In particular:".

So there is no way to know in general which of the two nodes is referred to here, which seems wrong.

Using `DT_INST(0, ...)` in any situation except a single-instance driver file is usually the wrong thing to do and it in general makes the code harder to reason about in terms of how it relates to the devicetree, and I do want to reserve the flexibility to reorder instances in the future if it makes sense for DT API reasons, so I'm on the lookout for places with hard-coded numbers.

If you are looking for a particular node, can you please use DT_NODELABEL or some other way to get a node identifier for it instead, to comply with the existing API rules, stay future proof, and keep this code readable? If not, why not?

danieldegrasse replied:
While the `usb_dc_mcux` driver file is single-instance, you are right that this configuration is wrong. I've updated the configuration to handle the `usb1` or `usb2` nodes being enabled, and in turn updated the USB driver to handle either node being enabled. The driver is still single-instance, but it now can correctly handle `usb2` being the node that is enabled, where previously only `usb1` would work.

Since this issue also existed on the RT10xx SOCs, I've updated the PR to enable `usb2` on those parts as well, and changed the title of the PR to reflect the broader set of changes.

danieldegrasse committed:
soc: rt10xx: Enable USB2 controller clock

Enable USB2 controller clock on RT10xx SOC init, if nodelabel has
status="okay"

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: usb_dc_mcux: Enable second USB EHCI

Enable second usb EHCI if usb2 node has status="okay" Note that this
driver is still an single instance driver, this change simply enables
the driver to work with the usb2 peripheral if that one is enabled, and
usb1 is disabled.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented:
Thanks @danieldegrasse !
